### PR TITLE
Wrong logic in transactions verify process - Closes #1928

### LINKED ISSUE #1928: Wrong logic in transactions verify process
### Expected behavior
We should check if transaction ID already exists in database earlier to avoid other checks (more expensive) to be performed.
### Actual behavior
Currently we check if transaction ID exists in database as the very last step (after all verifications/validations). It's not good to perform all those steps and then discard the transaction because we already have it.
https://github.com/LiskHQ/lisk/blob/982b858a195f248e1e21c298968f9a1de5d74943/logic/transaction.js#L679
### Steps to reproduce
N/A
### Which version(s) does this affect? (Environment, OS, etc...)
1.0.0

### PR DESCRIPTION
**Pull request by ManuGowda**
### What was the problem?
Currently, the transaction existed check is performed after all the validations, and the transaction is rejected later which is resulting in unwanted validation checks. It's invalid flow to perform all these checks steps and then discard the transaction because we already have it.
### How did I fix it?
As part of this PR, the check confirmed the transaction is performed first and then the validations are performed to check if the transaction is valid or not.
### How to test it?
Run `functional post` test.
### Review checklist

* The PR solves #1928 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: logic/transaction.js [hunk 1]
```diff
@@ -409,277 +409,285 @@ class Transaction {
 	 * @todo Add description for the params
 	 */
 	verify(transaction, sender, requester, cb, tx) {
-		let valid = false;
-		let err = null;
-
 		if (typeof requester === 'function') {
 			cb = requester;
 		}
```
Comment (nazarhussain) [inline logic/transaction.js]: @ManuGowda We are removing such practices of mutating function parameters, wherever we refactor the code. So please remove this if block and call the `verify` with the exact number of params.

File: test/functional/http/post/4.multisignature.advanced.js [hunk 2]
```diff
@@ -402,4 +402,17 @@ describe('POST /api/transactions (type 4) register multisignature', () => {
 			pendingMultisignatures
 		);
 	});
+
+	describe('verification', () => {
+		it('sending already confirmed second secret should fail', () => {
```
Comment (nazarhussain) [inline test/functional/http/post/4.multisignature.advanced.js]: The message is identical to test case above `second secret`. I believe its a typo mistake. It should be multisignature transaction.

File: test/functional/http/post/5.dapps.js [hunk 3]
```diff
@@ -710,4 +710,17 @@ describe('POST /api/transactions (type 5) register dapp', () => {
 	describe('confirmation', () => {
 		phases.confirmation(goodTransactions, badTransactions);
 	});
+
+	describe('verification', () => {
+		it('sending already confirmed multisignature transaction should fail', () => {
```
Comment (nazarhussain) [inline test/functional/http/post/5.dapps.js]: Same here.

File: test/functional/system/transactions/2.delegates/3.different.accounts.same.username.js [hunk 4]
```diff
@@ -136,7 +136,9 @@ describe('system test (type 2) - double delegate registrations', () => {
 
 					it('adding to pool delegate registration from same account should fail', done => {
 						localCommon.addTransaction(library, transaction2, err => {
-							expect(err).to.equal('Account is already a delegate');
+							expect(err).to.equal(
+								`Transaction is already confirmed: ${transaction2.id}`
+							);
```
Comment (nazarhussain) [inline test/functional/system/transactions/2.delegates/3.different.accounts.same.username.js]: The test case here 

```
adding to pool delegate registration from same account should fail
```

Does not mean that we will use same transaction. It means we will use same account. Either update the test description or expectation.

File: test/functional/system/transactions/2.delegates/4.different.accounts.different.usernames.js [hunk 5]
```diff
@@ -128,14 +128,18 @@ describe('system test (type 2) - double delegate registrations', () => {
 
 					it('adding to pool delegate registration with already registered username should fail', done => {
 						localCommon.addTransaction(library, transaction1, err => {
-							expect(err).to.equal('Account is already a delegate');
+							expect(err).to.equal(
+								`Transaction is already confirmed: ${transaction1.id}`
+							);
```
Comment (nazarhussain) [inline test/functional/system/transactions/2.delegates/4.different.accounts.different.usernames.js]: Same here.

File: test/functional/system/transactions/2.delegates/4.different.accounts.different.usernames.js [hunk 6]
```diff
@@ -128,14 +128,18 @@ describe('system test (type 2) - double delegate registrations', () => {
...
 							done();
 						});
 					});
 
 					it('adding to pool delegate registration from same account should fail', done => {
 						localCommon.addTransaction(library, transaction2, err => {
-							expect(err).to.equal('Account is already a delegate');
+							expect(err).to.equal(
+								`Transaction is already confirmed: ${transaction2.id}`
+							);
```
Comment (nazarhussain) [inline test/functional/system/transactions/2.delegates/4.different.accounts.different.usernames.js]: Same here.

File: test/functional/system/transactions/3.3.votes.js [hunk 7]
```diff
@@ -191,11 +191,7 @@ describe('system test (type 3) - voting with duplicate submissions', () => {
 
 					it('adding to pool downvoting transaction to same delegate from same account should fail', done => {
 						localCommon.addTransaction(library, transaction4, err => {
-							expect(err).to.equal(
-								`Failed to remove vote, delegate "${
-									accountFixtures.existingDelegate.delegateName
-								}" was not voted for`
-							);
+							expect(err).to.include('Transaction is already confirmed');
```
Comment (nazarhussain) [inline test/functional/system/transactions/3.3.votes.js]: Test scenario
`adding to pool downvoting transaction to same delegate from same account should fail`
is totally contradict to expectation. While the pervious exception was exact match to test statement.

File: test/functional/system/transactions/4.multisignature/4.4.multisignature.js [hunk 8]
```diff
@@ -130,7 +130,7 @@ describe('system test (type 4) - double multisignature registrations', () => {
 				library,
 				scenarios.regular.multiSigTransaction,
 				err => {
-					expect(err).to.equal('Account already has multisignatures enabled');
+					expect(err).to.have.string('Transaction is already confirmed');
```
Comment (nazarhussain) [inline test/functional/system/transactions/4.multisignature/4.4.multisignature.js]: Same here as above.

File: test/functional/system/transactions/4.multisignature/4.X.multisignature_validated.js [hunk 9]
```diff
@@ -101,7 +101,7 @@ describe('system test (type 4) - checking registered multisignature transaction
 				library,
 				scenarios.regular.multiSigTransaction,
 				err => {
-					expect(err).to.equal('Account already has multisignatures enabled');
+					expect(err).to.have.string('Transaction is already confirmed');
```
Comment (nazarhussain) [inline test/functional/system/transactions/4.multisignature/4.X.multisignature_validated.js]: Same here.

File: test/functional/system/transactions/2.delegates/4.different.accounts.different.usernames.js [hunk 10]
```diff
@@ -126,22 +126,34 @@ describe('system test (type 2) - double delegate registrations', () => {
...
-							expect(err).to.equal(
-								`Transaction is already confirmed: ${transaction2.id}`
-							);
-							done();
-						});
+						setTimeout(() => {
+							const transaction4 = lisk.transaction.registerDelegate({
+								passphrase: account2.password,
+								username: account2.username,
+							});
```
Comment (nazarhussain) [inline test/functional/system/transactions/2.delegates/4.different.accounts.different.usernames.js]: @ManuGowda Can't use we `timeOffset` here?

Comment (ManuGowda) [reply test/functional/system/transactions/2.delegates/4.different.accounts.different.usernames.js]: @nazarhussain sure we can use, but it's optional I believe.

File: test/functional/http/post/4.multisignature.js [hunk 11]
```diff
@@ -728,4 +728,17 @@ describe('POST /api/transactions (type 4) register multisignature', () => {
 			pendingMultisignatures
 		);
 	});
+
+	describe('verification', () => {
```
Comment (4miners) [inline test/functional/http/post/4.multisignature.js]: What is the difference between this tests and one included in `4.multisignature.advanced.js` file?

File: test/functional/http/post/1.second_secret.js [hunk 12]
```diff
@@ -144,4 +146,17 @@ describe('POST /api/transactions (type 1) register second secret', () => {
 	describe('confirmation', () => {
 		phases.confirmation(goodTransactions, badTransactions);
 	});
+
+	describe('verification', () => {
+		it('sending already confirmed second secret should fail', () => {
```
Comment (4miners) [inline test/functional/http/post/1.second_secret.js]: We're not sending `second secret`

File: test/functional/http/post/transactions.js [hunk 13]
```diff
@@ -43,4 +56,24 @@ describe('POST /api/transactions (general)', () => {
...
+
+	describe('transaction processing', () => {
+		sendTransactionPromise(transaction).then(res => {
+			expect(res.body.data.message).to.be.equal('Transaction(s) accepted');
+		});
+		return phases.confirmation([transaction], []);
+	});
+
+	describe('validation', () => {
+		it('should fail when an already confirmed transaction sent again', () => {
```
Comment (4miners) [inline test/functional/http/post/transactions.js]: not best description, maybe `should fail when trying to send already confirmed transaction / transaction that is already confirmed`

File: test/functional/http/post/4.multisignature.advanced.js [hunk 14]
```diff
@@ -402,4 +402,17 @@ describe('POST /api/transactions (type 4) register multisignature', () => {
 			pendingMultisignatures
 		);
 	});
+
+	describe('verification', () => {
+		it('sending already confirmed multi-signature should fail', () => {
```
Comment (4miners) [inline test/functional/http/post/4.multisignature.advanced.js]: conbsistency - `multisignature`, but what multisignature? registration, spend from multisignature account, etc? Please be precise.

File: test/functional/system/transactions/4.multisignature/4.4.multisignature.js [hunk 15]
```diff
@@ -126,14 +126,19 @@ describe('system test (type 4) - double multisignature registrations', () => {
 		});
 
 		it('adding to pool multisignature registration for same account should fail', done => {
-			localCommon.addTransaction(
-				library,
-				scenarios.regular.multiSigTransaction,
-				err => {
-					expect(err).to.equal('Account already has multisignatures enabled');
-					done();
+			const multiSignatureToSameAccount = lisk.transaction.registerMultisignature(
```
Comment (4miners) [inline test/functional/system/transactions/4.multisignature/4.4.multisignature.js]: Why that change was needed?

File: test/functional/http/post/transactions.js [hunk 16]
```diff
@@ -43,4 +56,24 @@ describe('POST /api/transactions (general)', () => {
...
 	});
+
+	describe('transaction processing', () => {
+		sendTransactionPromise(transaction).then(res => {
+			expect(res.body.data.message).to.be.equal('Transaction(s) accepted');
+		});
+		return phases.confirmation([transaction], []);
+	});
+
+	describe('validation', () => {
```
Comment (4miners) [inline test/functional/http/post/transactions.js]: consistency - `verification` or `validation`, globally

Comment (ManuGowda) [reply test/functional/http/post/transactions.js]: Thanks, @4miners. Actually, @nazarhussain suggest to remove all the individual tests and keep only verification test here`test/functional/http/post/transactions.js`, I will remove the repetitive tests.

File: logic/transaction.js [hunk 17]
```diff
@@ -409,277 +409,283 @@ class Transaction {
...
+				);
+			}
 
-		// Check timestamp
-		if (slots.getSlotNumber(transaction.timestamp) > slots.getSlotNumber()) {
-			return setImmediate(
-				cb,
-				'Invalid transaction timestamp. Timestamp is in the future'
+			// Call verify on transaction type
+			__private.types[transaction.type].verify.call(
```
Comment (4miners) [inline logic/transaction.js]: `this.checkConfirmed` should be called right before this line.

**Comment by MaciejBaj** [comment]
@ManuGowda please confirm that checking if a transaction is confirmed is more expensive than the verify process.

**Comment by ManuGowda** [comment]
@MaciejBaj here is the performance test details
```
Before
verify: 1.895ms
verify: 1.187ms
verify: 1.111ms
verify: 1.032ms
verify: 1.444ms
verify: 1.100ms
verify: 1.076ms
verify: 1.250ms
verify: 2.235ms
verify: 1.186ms
verify: 1.011ms
verify: 1.136ms
verify: 1.185ms
verify: 1.153ms
verify: 1.323ms
verify: 1.537ms
verify: 1.125ms
verify: 1.622ms
verify: 1.279ms
verify: 1.533ms
verify: 1.121ms
verify: 1.993ms
verify: 1.117ms
verify: 1.656ms
verify: 1.544ms
verify: 1.021ms
verify: 2.292ms
verify: 1.072ms
verify: 2.689ms
verify: 1.167ms
verify: 1.569ms
verify: 1.629ms
verify: 1.973ms
verify: 1.693ms
verify: 1.606ms
verify: 1.747ms
verify: 1.250ms
verify: 1.530ms
verify: 1.246ms
verify: 1.338ms
verify: 1.315ms
verify: 0.992ms
verify: 1.411ms
verify: 1.099ms
verify: 1.071ms
verify: 1.430ms
verify: 1.160ms
verify: 1.155ms
verify: 1.213ms
verify: 1.345ms
verify: 1.189ms
verify: 1.134ms
verify: 1.466ms
verify: 1.808ms
verify: 1.027ms
verify: 1.127ms
verify: 1.239ms
verify: 1.257ms
verify: 1.015ms
verify: 1.115ms
verify: 0.937ms
verify: 1.058ms
verify: 0.983ms
verify: 0.985ms
verify: 1.048ms
verify: 1.087ms
verify: 1.091ms
verify: 1.025ms
verify: 1.405ms
verify: 0.986ms
verify: 0.995ms
verify: 0.979ms
verify: 3.014ms
verify: 1.013ms
verify: 2.457ms
verify: 1.200ms
verify: 1.096ms
verify: 1.669ms
verify: 2.112ms
verify: 1.513ms
verify: 1.153ms
```

```
After
verify: 1.915ms         
verify: 1.112ms
verify: 1.105ms
verify: 1.215ms
verify: 1.084ms
verify: 1.225ms
verify: 1.323ms
verify: 1.227ms
verify: 1.160ms
verify: 1.053ms
verify: 1.006ms
verify: 1.010ms
verify: 0.958ms
verify: 1.445ms
verify: 1.121ms
verify: 1.397ms
verify: 1.011ms
verify: 1.534ms
verify: 0.981ms
verify: 1.205ms
verify: 2.410ms
verify: 3.652ms
verify: 0.954ms
verify: 1.179ms
verify: 1.144ms
verify: 1.155ms
verify: 1.026ms
verify: 2.782ms
verify: 1.712ms
verify: 1.024ms
verify: 1.781ms
verify: 1.097ms
verify: 1.119ms
verify: 1.504ms
verify: 1.895ms
verify: 1.491ms
verify: 1.283ms
verify: 1.065ms
verify: 1.107ms
verify: 1.259ms
verify: 1.538ms
verify: 1.310ms
verify: 1.526ms
verify: 1.074ms
verify: 1.163ms
verify: 1.046ms
verify: 1.427ms
verify: 1.256ms
verify: 0.939ms
verify: 1.053ms
verify: 1.334ms
verify: 2.014ms
verify: 1.117ms
verify: 1.440ms
verify: 2.373ms
verify: 1.141ms
verify: 1.350ms
verify: 2.724ms
verify: 1.016ms
verify: 1.377ms
verify: 1.615ms
verify: 1.346ms
verify: 0.956ms
verify: 0.896ms
verify: 0.989ms
verify: 2.186ms
verify: 1.070ms
verify: 1.201ms
verify: 1.250ms
verify: 1.866ms
verify: 1.224ms
verify: 1.159ms
verify: 1.142ms
verify: 1.159ms
verify: 1.263ms
verify: 2.324ms
verify: 1.144ms
verify: 1.118ms
verify: 1.180ms
verify: 1.185ms
verify: 1.859ms
verify: 1.749ms
```

**Pull request merged**